dashboard/head.py

```diff
@@ -74,6 +76,22 @@ async def _update_nodes(self):
         while True:
             try:
                 nodes = await self._get_nodes()
+
+                # Get correct node info by state,
+                #   1. The node is ALIVE if any ALIVE node info
```

Comment:
Can you explain the reason in the comment too?

Reply:
`GetAllNodeInfo` returns all nodes info including dead nodes. We needs to know how to merge them to ip. For example, there are two nodes info:
- {node id: 1, hostname: example}
- {node id: 2, hostname: example}

We choose which one for host example?

For one hostname, there will be a list of node info. and only two cases:
- All nodes info of one hostname are DEAD
- Only one node info of one hostname is ALIVE

So, here is the rule,
- Choose a DEAD one if all nodes info of one hostname are DEAD. 
- Choose the ALIVE one if there is ALIVE node in one hostname.

Reply:
It's better to add a timestamp to `GcsNodeInfo`.

Reply:
Hmm.. I see. So this is due to the case where there could be multiple raylets on a single host. I think this should be the case only when we use `cluster_utils` right? Or do you know any other case? I think it kind of doesn't make sense to have multiple ray start in a single node. 

If this is useful only for the cluster utils, why don't we just group by hostname + node_id and display both in the dashboard? Is there any issue with this?

Reply:
If a node failover many times, there will be a lot node info belongs to one hostname.

Reply:
I see. This hostname + node id idea still should work right? (We can just ignore all DEAD nodes). Is there any concern about this approach?

Reply:
If a node failover frequently, `hostname + node id` always get a DEAD node info. I think it's better to query the node info by hostname and show the history by DEAD node info. For example,

- {node id: 1, start_timestamp: 10, dead timestamp: 100, state: DEAD, hostname: example}
- {node id: 2, start_timestamp: 102, dead timestamp: 120, state: DEAD, hostname: example}
- {node id: 3, start_timestamp: 122, dead timestamp: 0, state: ALIVE, hostname: example}

then the node `example` will show the third node info with history:
- node id: 2, start at: 102, dead at: 120
- node id: 1 start at: 10, dead at: 100

Reply:
> If a node failover frequently, hostname + node id always get a DEAD node info.
In this case, we can just not render them right? We can use your logic in the frontend.

The reason why I want to use these as keys is because it'll be great if we can work on dashboard with cluster_utils. The current structure won't work with cluster_utils because it will have the same hostname for multiple node ids.

Reply:
@mfitton @rkooo567 We want to find the best solution between (IP, hostname) and (node id).

- For (IP, hostname)
  - Pros: simple for users, friendly to one host one node. (It's easy to show the host history)
  - Cons: not support one host multiple nodes, so it's not work well with cluster_utils.

- For (node id)
  - Pros: support one host multiple nodes, work well with cluster_utils.
  - Cons: too many nodes info may confuse users, hard to find out what happened if node failover many times.

Reply:
I think we can hide the second options' con easily by filtering DEAD node ids to user facing APIs right?

Reply:
{
node_id: {
  ip:
  host:
  state:
}
}

And in the frontend, we can just filter DEAD state node id.

Reply:
@rkooo567 I see. Use node id as the key of node info will returns full nodes info to front-end. And front-end do the filter job. My concern is that the data will be too large if we run many nodes in one cluster. For the DEAD node, the physics node info is useless. If we choose `node id` as the key, we need some strategies to reduce the data size:
- Only contains `GcsNodeInfo` data for the DEAD nodes.
- Each (IP + hostname) reserves a limited number of DEAD nodes. For example, the last 5 DEAD nodes.

Reply:
I see your concerns (since you guys are running long-running clusters, it can have lots of information). I like the second solution, but we can allow users to configure them. So, for regular usage, it is 0~2, and for cluster_utils cases, it can be like 10. 

> Only contains GcsNodeInfo data for the DEAD nodes.

For this, I am probably not familiar how the data is represented now (because I thought this was the current behavior). Can you explain a little bit?

Reply:
Current node info is a mixture of `node physical stats` (from **_reporter agent_**) and `node stats` (from **_GetNodeStats_** rpc). If a node is dead, the `node physical stats` and `node stats` will be unreliable, only `GcsNodeInfo` is correct.